master/internal/task/allocation.go

```diff
@@ -1093,63 +1117,87 @@ func (a *allocation) terminated(reason string) {
 	case a.killedWhileRunning:
 		exitReason = fmt.Sprintf("allocation killed after %s", reason)
 		a.syslog.Info(exitReason)
+		a.SetExitStatus(exitReason, "", nil)
 		return
 	case a.req.Preemptible && preemptible.Acknowledged(a.req.AllocationID.String()):
 		exitReason = fmt.Sprintf("allocation preempted after %s", reason)
 		a.syslog.Info(exitReason)
+		switch err := a.exitErr.(type) {
+		case sproto.ResourcesFailure:
```

Comment:
weirdly, resources failure is one of the only ones that shouldn't have an err code (it should be called "ResourcesRestoreFailure" and only has to do with restore failures).

i think i can help save you a lot of code here using `defer`. let's pair for like 15m on this tomorrow if that is ok.